File: Civitas-Java/Civitas/src/main/java/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 * @author elena
 * @date 26/09/2019
 */

import java.util.ArrayList;
import java.util.Random;

public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    Diario diariomazo = Diario.getInstance();
    
    private void init(){
        sorpresas = new ArrayList();
        cartasEspeciales = new ArrayList();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(boolean debug){
        this.debug = debug;
        init();       
        
        if(debug)
            diariomazo.ocurreEvento("Debug");
    }
    
    MazoSorpresas(){
        init();
        this.debug = false;
    }
    
    void alMazo(Sorpresa s){
        if(!barajada)
            sorpresas.add(s);        
    }
    
    Sorpresa siguiente(){
        if(!barajada || usadas == sorpresas.size()){
            if(!debug){
                barajar();
                usadas = 0;
                barajada = true;
            }                
        }
        
        usadas++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(ultimaSorpresa);
        
        return ultimaSorpresa;
    }
    
    void barajar(){   
        Random rand = new Random();

        for(int i = 0; i < sorpresas.size(); i++){ 
            int segunda = rand.nextInt(sorpresas.size());

            Sorpresa temp = sorpresas.get(i);
            sorpresas.set(i, sorpresas.get(segunda));
            sorpresas.set(segunda, temp);
        } 
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        for(int i = 0; i < sorpresas.size(); i++){
            if(sorpresa == sorpresas.get(i)){
                sorpresas.remove(i);
                cartasEspeciales.add(sorpresa);
                diariomazo.ocurreEvento("Inhabilitar carta especial");
            }
        }
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        for(int i = 0; i < cartasEspeciales.size(); i++){
            if(sorpresa == cartasEspeciales.get(i)){
                cartasEspeciales.remove(i);
                sorpresas.add(sorpresa);
                diariomazo.ocurreEvento("Habilitar carta especial");
            }
        }
    }
}

File: Civitas-Ruby/lib/mazo_sorpresas.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "sorpresa"
require_relative "diario"

module Civitas
  class MazoSorpresas
    def initialize
      @sorpresas = Array.new()
      @barajada = false
      @usadas = 0
      @debug = false
      @cartasEspeciales = Array.new()
      @ultimaSorpresa = nil
      @diariomazo = Diario.instance      
    end
    
    attr_accessor :debug
    
    def almazo(s)
      if !@barajada
        @sorpresas << s
      end
    end
    
    def siguiente
      if !@barajada or @usadas.eql?(@sorpresas.length)
        if !@debug
          @sorpresas.shuffle
          @usadas = 0
          @barajada = true
        end
      end
      
      @usadas = @usadas + 1
      @ultimaSorpresa = @sorpresas[0]
      @sorpresas.delete_at(0)
      @sorpresas << @ultimaSorpresa
      
      return @ultimaSorpresa
    end
    
    def inhabilitarcartaespecial(sorpresa)
      for s in @sorpresas do
        if s.nombre == sorpresa.nombre
          @sorpresas.delete(sorpresa)
          @cartasEspeciales << sorpresa
          @diariomazo.ocurre_evento("Inhabilitar carta especial")
        end
      end
    end
    
    def habilitarcartaespecial(sorpresa)
      for s in @cartasEspeciales do
        if s.nombre == sorpresa.nombre
          @cartasEspeciales.delete(sorpresa)
          @sorpresas << sorpresa
          @diariomazo.ocurre_evento("Habilitar carta especial")
        end
      end
    end
  end
end
